File: cpp/rotational-cipher/rotational_cipher.cpp
#include "rotational_cipher.h"

namespace rotational_cipher {

bool is_char(char c) {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
}

bool is_lowercase(char c) {
    return c >= 'a' && c <= 'z';
}

char encode_char(char c, int rot) {
    if (is_char(c)) {
        if (is_lowercase(c)) {
            return ((c + rot - 'a') % 26) + 'a';
        } else {
            return ((c + rot - 'A') % 26) + 'A';
        }
    } else {
        return c;
    }
}

string rotate(string input, int rot) {
    string encoded = "";

    for (unsigned int i = 0; i < input.length(); ++i) {
        encoded.push_back(encode_char(input.at(i), rot));
    }

    return encoded;
}

}  // namespace rotational_cipher

File: go/rotational-cipher/rotational_cipher.go
package rotationalcipher

func isUpperCase(c byte) bool {
	return c >= 'A' && c <= 'Z'
}

func isLowerCase(c byte) bool {
	return c >= 'a' && c <= 'z'
}

func encodeChar(c byte, shiftKey int) byte {
	if isUpperCase(c) {
		charOrd := (c + byte(shiftKey) - byte('A')) % 26
		return byte('A') + charOrd
	} else if isLowerCase(c) {
		charOrd := (c + byte(shiftKey) - byte('a')) % 26
		return byte('a') + charOrd
	} else {
		return c
	}
}

func RotationalCipher(plain string, shiftKey int) string {
	if shiftKey < 0 || shiftKey >= 27 {
		panic("invalid key")
	}

	encoded := make([]byte, len(plain))

	for i := 0; i < len(plain); i++ {
		encoded[i] = encodeChar(plain[i], shiftKey)
	}

	return string(encoded)
}
